# Pacotes

Pacotes são coleções de funções, códigos e ferramentas para implementar tarefas de forma fácil no Python. Uma vez que os baixamos com um gerenciador de pacotes (`conda` ou `pip`), basta importá-los no nosso código.

# Numpy

Numpy é um pacote Python feito para lidar com números, vetores, matrizes e álgebra no geral, com alta performance e facilidade.

In [8]:
#Antes de qualquer coisa, vamos importar o NumPy no nosso codigo como  uma abreviação 'np'
import numpy as np #IMPORTAR numpy COMO np


Agora já podemos fazer uso de todas as funcionalidades do `numpy` no nosso código! Note que podemos importar o `numpy` como qualquer abreviação, e até mesmo sem uma! Por exemplo:

>```Python
import numpy
>```

No entanto, se quisermos usar a função `array` do `numpy`, por exemplo, precisaremos digitar `numpy.array` e assim será para todas as funcionalidades. Portanto, é conveniente utilizar uma abreviação, neste caso digitaremos apenas `np.array`.

**É uma convenção utilizar `np` como abreviação para `numpy`.**

## tipos Numpy

Para uma lista completa, atualizada e explicada de todos os tipos de variáveis básicas do `NumPy` (em inglês) [clique aqui](https://numpy.org/devdocs/user/basics.types.html).

Vamos ver os tipos principais básicos que utilizaremos no Numpy. Isso servirá para especificar a precisão de uma variável.

### signed ints

Um `signed int` é um inteiro que utiliza um dos bits para guardar o sinal do número.

| tipo Numpy | tipo C | bits | Descrição | nomes alternativos (bits) |
| :- | :- | :- | :-| :- |
| `np.byte` | char | 8  | Definido pela plataforma.| `np.int8`|
| `np.short`| short | 16  | Definido pela plataforma.| `np.int16`|
| `np.intc` | int | 32  | Definido pela plataforma.| `np.int32`|
| `np.int_`| long | 64  |Definido pela plataforma.| `np.int64`|

### unsigned ints

Um `unsigned int` é um  inteiro que não utilizada nenhum bit para guardar o sinal do número e, portanto, é sempre positivo e tem um alcance maior.

| tipo Numpy | tipo C | bits | Descrição | nomes alternativos (bits) |
| :- | :- | :- | :-| :- |
| `np.ubyte` | char | 8 | Definido pela plataforma.| `np.uint8`|
| `np.ushort`| short | 16 | Definido pela plataforma.| `np.uint16`|
| `np.uintc` | int | 32  | Definido pela plataforma.| `np.uint32`|
| `np.uint`| long | 64  |Definido pela plataforma.| `np.uint64`|

| tipo Numpy |nomes alternativos| tipo C | bits | Descrição  | nomes alternativos (bits) |
| :- | :- | :-: | :-: | :- | :-|
|`np.half`| - |-| 16  | Float de meia precisão: expoente de 5 bits e mantissa de 10 bits.|`np.float16`|
| `np.single`|-| float |  32  | Float de precisão única definido pela plataforma: normalmente um signed bit com expoente de 8 bits e uma mantissa de 23 bits.|  `np.float32` |
| `np.double`|`np.float_`| double | 64  | Float de precisão dupla definido pela plataforma: normalmente um signed bit com expoente de 11 bits e uma mantissa de 52 bits.|`np.float64`|
| `np.longdouble`| `np.longfloat`| long double |128  | Float de precisão estendida definido pela plataforma: normalmente um signed bit com expoente de 11 bits e uma mantissa de 52 bits.| **`np.float128`** \*|
| `np.csingle`|`np.singlecomplex`| float complex |2x32 | Número complexo representado por 2 floats de precisão única (parte real e imaginária).| `np.complex64 `|
| `np.cdouble`|`numpy.cfloat`, `numpy.complex_`| double complex  | 2x64 | Número complexo representado por 2 floats de precisão dupla (parte real e imaginária).|`np.complex128`|
| `np.clongdouble`|`numpy.clongfloat`, `numpy.longcomplex`| long double complex | 2x128 | Número complexo representado por 2 floats de precisão estendida (parte real e imaginária).| **`np.complex256`** \*|


A última coluna mostra alternativas com o número de bits explícitos. Items em negrito, marcados com asterisco, indicam nomenclaturas alternativas que podem não existir em algumas plataformas, vamos ver como checar isso usando a função `info` mais tarde.

In [2]:
import numpy as np

print('\n Reais:')
print('np.float16:    ', np.float16(2/3))
print('np.float32:    ', np.float32(2/3))
print('np.float64:    ', np.float64(2/3))
print('np.longdouble: ', (2/np.longdouble(3)))

print('\n Complexos:')
print('np.complex64:   ', np.complex64(2/3 + 1j/3))
print('np.complex128:  ', np.complex128(2/3 + 1j/3))
print('np.clongdouble: ', np.clongdouble(2 + 1j)/3)


 Reais:
np.float16:     0.6665
np.float32:     0.6666667
np.float64:     0.6666666666666666
np.longdouble:  0.6666666666666666667

 Complexos:
np.complex64:    (0.6666667+0.33333334j)
np.complex128:   (0.6666666666666666+0.3333333333333333j)
np.clongdouble:  (0.6666666666666666667+0.33333333333333333334j)


Note que o `numpy` avalia a operação dentro da função de conversão antes de executá-la, então, por exemplo

* `np.longdouble(2/3)` criará um `np.longdouble` que não necessariamente corresponderá a `2/3` até a última casa decimal disponível. Isso porque o Python fará nessa ordem:
    1. `2/3 = 0.6666666666666666` $\rightarrow$ salva-se `2/3` como um `np.float64`.
    2. `np.longdouble(0.6666666666666666)` $\rightarrow$ transforma-se um `np.float64` em um `np.longdouble`.
    3. As casas decimais extras são adicionadas sem conhecimento de que `0.6666666666666666` é uma aproximação de `2/3`.
    

Solução: 

* Define-se a fração definindo um dos números como `np.longdouble` e então executa-se as operações. Veja uma comparação:

    1. `np.longdouble(2)/3 = 0.6666666666666666667`
    2. `2/np.longdouble(3) = 0.6666666666666666667`
    3. `np.longdouble(2/3) = 0.66666666666666662966`   
    
Uma vez que uma variável `np.longdouble` foi definida, qualquer operação com ela resultará em outra `np.longdouble`.

### info()

[Clique aqui](https://numpy.org/doc/stable/reference/generated/numpy.info.html) para abrir a documentação mais recente sobre a função (em inglês).

Nós podemos usar a função `info` do `numpy` para exibir informações específicas de um tipo de variável para nossa plataforma:

Input:
>```Python
import numpy as np #importa numpy como np
print(np.info(np.int8)) #printa informações sobre o tipo 'np.int8' na minha plataforma
>```

Output:
>int8()
>
>Signed integer type, compatible with C ``char``.
>
>:Character code: ``'b'``
>
>:Canonical name: `numpy.byte`
>
>:Alias on this platform: `numpy.int8`: 8-bit signed integer (``-128`` to ``127``).
>
>Methods:...

Legenda:

* **Character code**: o código para esse tipo de variável.
* **Canonical name**: o nome padrão para esse tipo de variável.
* **Alias on this platform**: nomes alternativos para esse tipo de variável e informações sobre sua precisão na sua plataforma.
* **Methods**: lista de todos os métodos disponíveis para essa variável.

### finfo()

[Clique aqui](https://numpy.org/doc/stable/reference/generated/numpy.finfo.html) para abrir a documentação mais recente sobre a função (em inglês).

Alternativamente, podemos usar `finfo` para exibir apenas informações detalhadas sobre a precisão de um tipo de float (ou complexo) na nossa máquina. Por exemplo, para exibir informações do tipo `np.double`, podemos printar o resultado da função `finfo` nesse tipo:

In [7]:
import numpy as np
print(np.finfo(np.float64))

Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
---------------------------------------------------------------



| quantidade |tipo de valor| significado |
| :- |:-| :- |
|**bits**|int |O número de bits ocupado pelo tipo.|
|**eps**|float|A diferença entre 1.0 e o próximo menor float representável maior que 1.0. Por exemplo, para um float binário de 64 bits no padrão IEEE-754, eps=$2^{-52}$, aproximandamente $2.22\times 10^{-16}$.|
|**epsneg**|float | A diferença entre 1.0 e o próximo menor float representável maior que 1.0. Por exemplo, para um float binário de 64 bits no padrão IEEE-754, eps=$2^{-53}$, aproximandamente $1.11\times 10^{-16}$.|
|**iexp**|int|O número de bits na parte do expoente na representação do float.|
|**machar**|MachAr|O objeto que fez esses cálculos e tem mais informações detalhadas.|
|**machep**|int|O expoente que produz **eps**. |
|**max**|floating point number of the appropriate type| O maior número representável.|
|**maxexp**|int|O menor valor positivo para a potência em base (2) que causa overflow.|
|**min**|floating point number of the appropriate type|O menor valor representável, normalmente -**max**.|
|**minexp**|int|O valor mais negativo do potência em base (2) consistente com não haver 0's na frente da mantissa.|
|**negep**|int|O expoente que produz **epsneg**.|
|**nexp**|int|O número de bits no expoente incluindo o sinal e o bias.|
|**nmant**|int|O número de bits na mantissa.|
|**precision**|int|O número aproximado de casas decimais para qual esse tipo de float é preciso.|
|**resolution**|floating point number of the appropriate type|A resolução aproximada para este tipo, i.e., $10^{precision}$|
|**tiny**|float|O menor valor possível para um float com precisão total.|

### iinfo()

[Clique aqui](https://numpy.org/doc/stable/reference/generated/numpy.iinfo.html) para abrir a documentação mais recente sobre a função.

Um comando análogo ao `finfo()` porém para **inteiros**, o método de utilização é o mesmo:

In [2]:
import numpy as np
print(np.iinfo(np.byte))

Machine parameters for int8
---------------------------------------------------------------
min = -128
max = 127
---------------------------------------------------------------



## Arrays (arranjos)

[Clique aqui](https://numpy.org/devdocs/reference/arrays.html) para abrir a documentação mais atualizada sobre arrays do `NumPy` (em inglês).

Um arranjo no `numpy` é um 'contêiner' multidimensional (geralmente de tamanho fixo) de itens do mesmo tipo e tamanho.

<center><img style="float: center" src="files/nparray0.png" width="800"></center>


Todos os elementos de um arranjo multidimensional são:

   * do mesmo tipo.
   * do mesmo tamanho (em bits). 

Uma maneira de criar um array é usando a função `array` do `numpy`. **Lembrando que os parâmetros de palavra-chave (keyword arguments ou simplesment kwargs, aqueles parâmetros que possuem um `=` explícito) são opicionais**.  Vejamos as principais utilidades da função `array` para nós:

### np.array()

* `np.array(object, dtype=None, order='K', copy=True, subok=False, ndmin=0, like=None)`

    * `object`: um objeto do tipo arranjo (lista, tupla, etc)
    * `dtype`: tipo de variável dos elementos (`int`, `float`, `complex`, etc). Se omitido, será selecionado automaticamente o menor tipo necessário para guardar os elementos.
    * `order`: 'C', 'F', 'K' ou 'A'. O tipo de ordenação do arranjo:
        * 'C' = Ordenação da linguagem C, as linhas são o primeiro índice.
        * 'F' = Ordenação da linguagem Fortran, as colunas são o primeiro índice.
        * 'A' = Caso `object` seja um array, ordenação do Fortran se o input for F, caso contrário C.
        * 'K' = Caso `object` seja um array, ordenação do C e do Fortran são preservadas nos array originais, caso contrário, assume-se a 'ordenação mais similiar'.
        * Se omitido, assume-se `order='K'`
    * `copy`: booleana (`True` ou `False`). Caso `copy=True`, uma cópia do arranjo é criada na memória. Se omitido, assume-se `True`.

Vamos criar um arranjo unidimensional equivalente o seguinte vetor $a$ usando `np.array()`:

$$a=\begin{bmatrix}1 & 0 & 1 \end{bmatrix}$$

In [13]:
import numpy as np 
b=[1,0]
a=np.array(b)
print(a)

[1 0]


Vamos criar um arranjo bidimensional equivalente a matriz $B$ usando `np.array()`, com os elementos do tipo `np.float64`:

$$B=\begin{bmatrix}1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{bmatrix}$$

In [15]:
import numpy as np
a=[[1,0,0] , [0,1,0] , [0,0,1]]
B=np.array(a, dtype=np.float64)
print(B)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


### Métodos de informações sobre o arranjo

Para um arranjo `B`, podemos obter informações sobre seu layout na memória com os seguintes comandos:

|Método|Descrição| output para nosso arranjo `B` acima |
|:-|:-|:-|
|`B.shape`| Uma tupla com as dimensões do arranjo. | (3,3)|
|`B.ndim`| Número de dimensões do arranjo. | 2 |
|`B.size`| Número de elementos do arranjo. | 9 |
|`B.itemsize`| Tamanho de um elemento do arranjo, em bytes. | 8 |
|`B.nbytes`| Tamanho total consumido pelo arranjo, em bytes. | 72 |
|`B.dtype`| Tipo de dados armazenados no arranjo. | dtype('float64') |
|`B.flags`| Informações sobre o layout do arranjo na memória.|


In [19]:
print(B.shape)

(3, 3)


No geral, um arranjo bidimensional $A$ de tamanho $n\times m$ terá a forma:

$$A=\begin{bmatrix}a_{11} & a_{12} & ... & a_{1m} \\ a_{21} & a_{22} & ... & a_{2m} \\ \vdots & \vdots & \ddots & \vdots \\ a_{n1} & a_{n2} & ... & a_{nm}\end{bmatrix}$$

Vamos definir um arranjo $C$ da seguinte forma:

$$C=\begin{bmatrix}1 & 2 \\ 3 & 4 \end{bmatrix}$$

In [30]:
import numpy as np
C=np.array([[1,2],[3,4]])
print(C)

[[1 2]
 [3 4]]


### Métodos/funções do formato

Podemos usar algumas ferramentas para manipular o formato dos nossos arranjos. Como elas funcionam como funções, precisamos colocar `()` no final do comando para chamá-las.


|Método para um arranjo `C`| Descrição| Função equivalente + Link para documentação |
|:- |:- | :-|
|`C.reshape(shape[, order])`|Retorna um arranjo contendo um novo formato (shape).|[np.reshape()](https://numpy.org/devdocs/reference/generated/numpy.reshape.html)|
|`C.resize(new_shape[, refcheck])`|Modificar o tamanho e o formato do arranjo na hora.| [np.resize()](https://numpy.org/devdocs/reference/generated/numpy.resize.html)|
|`C.transpose(*axes)`|Retorna a visualização da transposta de um arranjo.|[np.tranpose()](https://numpy.org/devdocs/reference/generated/numpy.transpose.html)|
|`C.swapaxes(axis1, axis2)`|Retorna a visão de um arranjo com `axis1` e `axis2` trocados.|[swapaxes()](https://numpy.org/devdocs/reference/generated/numpy.swapaxes.html)|
|`C.flatten([order])`|Retorna uma cópia do arranjo colapsado em uma dimensão.|[Método apenas](https://numpy.org/devdocs/reference/generated/numpy.ndarray.flatten.html)|
|`C.ravel([order])`|Retorna um arranjo colapsado para uma dimensão.| [np.ravel()](https://numpy.org/devdocs/reference/generated/numpy.ravel.html)|
|`C.squeeze([axis])`|Remove eixos de tamanho 1 do arranjo.| [np.squeeze()](https://numpy.org/devdocs/reference/generated/numpy.squeeze.html)|

In [34]:
import numpy as np
C=np.array([[1,2],[3,4]])
reshapeC=np.reshape(C,(4,1))
resizeB=np.resize(C,(1,10))

print('C: ')
print(C)
print('reshape(4,1): ')
print(reshapeC)
print('resize(1,10): ')
print(resizeB)

C: 
[[1 2]
 [3 4]]
reshape(4,1): 
[[1]
 [2]
 [3]
 [4]]
resize(1,10): 
[[1 2]]


Os operadores básicos de aritmética (`*`,`/`,`//`,`**`, etc), também funcionam com arranjos, porém essas operações são feitas sempre **entre elementos**. 

* **A multiplicação matricial será dada pelo operador `@`.**


Vamos criar um arranjo para representar uma matriz identidade `I` (2x2) e operá-la com a matriz `C` que fizemos.

`I*C` fará uma operação de elemento por elemento.
`I@C` fará uma multiplicação matricial.

In [22]:
I=np.array([[1,0],[0,1]]) #'\n é o operador de quebra de linha para strings, ele fará com que o print
C=np.array([[1,1],[1,1]])
print('\n I: \n',I)       #pule para a próxima linha, no meio de uma string, na hora de escrever
print('\n C: \n',C)
print('\n I*C: \n ',I*C)
print('\n I@C: \n',I@C)


 I: 
 [[1 0]
 [0 1]]

 C: 
 [[1 1]
 [1 1]]

 I*C: 
  [[1 0]
 [0 1]]

 I@C: 
 [[1 1]
 [1 1]]


* Vários **métodos que alteram o arranjo** podem ser usadas ou como funções ou como métodos 
    * Cheque as tabelas acima parar se certificar que existe uma função equivalente ao método.
    

* Embora, em geral, a síntaxe do método seja mais simples, **dê preferência para a função**, já que alguns métodos alterarão seu arranjo na hora e outros não, entre outros detalhes. 
    * Na dúvida, consulte a documentação sobre o método específico.

* `reshape()` nos permitirá mudar o formato do nosso arranjo desde que o novo arranjo tenha o mesmo número de elementos que o antigo:

$$C=\begin{bmatrix}1 & 2 \\ 3 & 4 \end{bmatrix} \rightarrow \textbf{C=np.reshape(C, (4,1))}  \rightarrow C=\begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}$$

* `resize()` é similar a função `reshape()` mas nos permite mudar o tamanho do arranjo (o número de elementos).
    * Caso o número de elementos do novo arranjo seja menor que o original, ele será truncado.
    * Caso seja maior, por padrão, `resize()` o preencherá repetindo valores do arranjo original.

$$C=\begin{bmatrix}1 & 2 \\ 3 & 4 \end{bmatrix} \rightarrow \textbf{C=np.resize(C, (1,8))} \rightarrow C=\begin{bmatrix} 1 & 2 & 3&4&1&2&3&4 \end{bmatrix}$$

In [44]:
a=np.resize(C,(1,8))
print(a)

[[1 2 3 4 1 2 3 4]]


### Métodos/Funções de cálculo básicas

|Método para um arranjo `C`| Descrição| Função equivalente + Link para documentação |
|:- |:- | :-|
|`C.max([axis, out, keepdims, initial, …])`|Retorna o valor máximo ao longo de um eixo.|[np.amax()](https://numpy.org/devdocs/reference/generated/numpy.amax.html)|
|`C.argmax([axis, out])`|Retorna os índices do valor máximo ao longo de um eixo.| [np.argmax()](https://numpy.org/devdocs/reference/generated/numpy.argmax.html)|
|`C.min([axis, out, keepdims, initial, …])`|Retorna o valor mínimo ao longo de um eixo.| [np.amin()](https://numpy.org/devdocs/reference/generated/numpy.amin.html)|
|`C.argmin([axis, out])`|Retorna os índices do valor mínimo ao longo de um eixo.|[np.argmin()](https://numpy.org/devdocs/reference/generated/numpy.argmin.html)|
|`C.ptp([axis, out, keepdims])`|Diferença entre o máximo e o mínimo ao longo de um eixo.| [np.ptp()](https://numpy.org/devdocs/reference/generated/numpy.ptp.html)|
|`C.clip([min, max, out])`|Retorna um arranjo cujos valores são limitados entre [min, max].| [np.clip()](https://numpy.org/devdocs/reference/generated/numpy.clip.html)|
|`C.conj()`|Complexo conjugado de todos os elementos.| [np.conj()](https://numpy.org/devdocs/reference/generated/numpy.conj.html)|
|`C.round([decimals, out])`|Retorna a com cada elemento arredondado para um número de casas decimais.|[np.around()](https://numpy.org/devdocs/reference/generated/numpy.around.html)|
|`C.trace([offset, axis1, axis2, dtype, out])`|Retorna a soma ao longo das diagonais de um eixo.|[np.trace()](https://numpy.org/devdocs/reference/generated/numpy.trace.html)|
|`C.sum([axis, dtype, out, keepdims, …])`|Retorna a soma dos elementos ao longo de um eixo.|[np.sum()](https://numpy.org/devdocs/reference/generated/numpy.sum.html)|
|`C.cumsum([axis, dtype, out])`|Retorna a soma cumulativa dos elementos ao longo de um eixo.|[np.cumsum()](https://numpy.org/devdocs/reference/generated/numpy.cumsum.html)|
|`C.mean([axis, dtype, out, keepdims, where])`|Returns the average of the array elements along given axis.|[np.mean()](https://numpy.org/devdocs/reference/generated/numpy.mean.html)|
|`C.var([axis, dtype, out, ddof, …])`|Retorna a variância dos elementos ao longo de um eixo.|[np.var()](https://numpy.org/devdocs/reference/generated/numpy.var.html)|
|`C.std([axis, dtype, out, ddof, …])`|Retorna o erro padrão dos elementos ao longo de um eixo.|[np.std()](https://numpy.org/devdocs/reference/generated/numpy.std.html)|
|`C.prod([axis, dtype, out, keepdims, …])`|Retorna o produto dos elementos ao longo de um eixo.| [np.prod()](https://numpy.org/devdocs/reference/generated/numpy.prod.html)|
|`C.cumprod([axis, dtype, out])`|Retorna o produto cumulativo dos elementos ao longo de um eixo.| [np.cumprod()](https://numpy.org/devdocs/reference/generated/numpy.cumprod.html)|
|`C.all([axis, out, keepdims, where])`|Retorna True se todos os elementos forem True.|[np.all()](https://numpy.org/devdocs/reference/generated/numpy.all.html)|
|`C.any([axis, out, keepdims, where])`|Retorna True se qualquer um dos elementos a for True.|[np.any()](https://numpy.org/devdocs/reference/generated/numpy.any.html)|

## Criando intervalos numéricos (1D arrays)

Algumas funções do `numpy` similiar ao `range()` do Python nos permitirão criar arranjos de forma similar, porém com argumentos podendo ser reais ou inteiros.

### np.arange()

A função `np.arange()` é bem similar ao `range()` (até no nome), mas nos permitirá especificar mais coisas, além de retornar um objeto do tipo `array` ao invés de um objeto do tipo `range`. ([Documentação](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html#numpy.linspace))

`np.arange(start=0, stop, step=1, dtype=None)`
   * `start` é um `float` ou `int` opcional com valor padrão `0`
   * `stop` é um `float` ou `int` obrigatório.
   * `step` é um `float` out `int` opcional com valor padrão `1`.
   * `dtype` é um tipo de variável numérica opcional, caso omitido, o valor padrão será definido pelos elementos do `array`.
   * **Essa função retorna:** um `array` que representa um **intervalo semi-aberto** discreto $[start, \, stop)$, com elementos que vão de `start` até `stop` com passo `step`. Note que como o intervalo é semi-aberto, `stop` não será parte do `array`.
      * **Obs: para valores `step` não inteiros, os resultados normalmente não serão consistentes. Neste caso é melhor usar a função `linspace()` que já vamos falar sobre.**

In [46]:
a=np.arange(0,5,0.5)
print(a)

[0.  0.5 1.  1.5 2.  2.5 3.  3.5 4.  4.5]


### np.linspace()

A função `np.linspace()` é similar a `range()` e `np.arange()`, porém, com parâmetros diferentes. ([Documentação](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html#numpy.linspace))

` np.linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None, axis=0)`
   * `start` é um `float` ou `int` obrigatório.
   * `stop` é um `float` ou `int` obrigatório.
   * `num` é um `int` opcional com valor padrão `50`.
   * `endpoint` é um `boolean` com valor padrão `True`.
   * `retstep` é um `boolean` com valor padrão `False`.
   * `dtype` é um tipo de variável numérica opcional, caso omitido, o valor padrão será definido pelos elementos do `array`.
   * **Essa função retorna:** um `array` que representa um **intervalo fechado** discreto $[start, \, stop]$, com `num` elementos que vão de `start` até `stop`.`endpoint` diz se o intervalo precisa conter `stop`. `retstep` diz para a função retornar além do `array`, o espaçamento entre os elementos do arranjo (equivalente ao `step` do `np.arange()`.
      * **Obs: para valores `step` não inteiros, os resultados normalmente não serão consistentes. Neste caso é melhor usar a função `linspace()` que já vamos falar sobre.**

In [24]:
a, dx=np.linspace(-11, 11, num=15, retstep=True)
print(a,dx)

[-11.          -9.42857143  -7.85714286  -6.28571429  -4.71428571
  -3.14285714  -1.57142857   0.           1.57142857   3.14285714
   4.71428571   6.28571429   7.85714286   9.42857143  11.        ] 1.5714285714285714
